sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -201,26 +201,28 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
   }
 
   /**
-   * Returns whether an expression is likely to be selective
+   * Returns whether an expression is likely to be selective. If the filtering predicate is on join
+   * key then partition filter can be inferred statically in optimization phase, hence return false.
    */
-  private def isLikelySelective(e: Expression): Boolean = e match {
-    case Not(expr) => isLikelySelective(expr)
-    case And(l, r) => isLikelySelective(l) || isLikelySelective(r)
-    case Or(l, r) => isLikelySelective(l) && isLikelySelective(r)
-    case _: StringRegexExpression => true
-    case _: BinaryComparison => true
-    case _: In | _: InSet => true
-    case _: StringPredicate => true
-    case _: MultiLikeBase => true
+  private def isLikelySelective(e: Expression, joinKey: Expression): Boolean = e match {
+    case Not(expr) => isLikelySelective(expr, joinKey)
+    case And(l, r) => isLikelySelective(l, joinKey) || isLikelySelective(r, joinKey)
+    case Or(l, r) => isLikelySelective(l, joinKey) && isLikelySelective(r, joinKey)
+    case expr: StringRegexExpression => true && !expr.references.subsetOf(joinKey.references)
+    case expr: BinaryComparison => true && !expr.references.subsetOf(joinKey.references)
+    case expr: In => true && !expr.references.subsetOf(joinKey.references)
+    case expr: InSet => true && !expr.references.subsetOf(joinKey.references)
+    case expr: StringPredicate => true && !expr.references.subsetOf(joinKey.references)
+    case expr: MultiLikeBase => true && !expr.references.subsetOf(joinKey.references)
```

Comment:
I only was little concerning about if all these predicates can be supported by the static partition pruning path. but yea by no means this should be duplicated with static path.

Reply:
@viirya so is it safe to say predicates defined in this method are the supported ones? org.apache.spark.sql.execution.datasources.DataSourceStrategy#translateLeafNodeFilter

If can just make y change for supported filters.

Reply:
`translateLeafNodeFilter` is used for data filtering pushdown. Seems for the static partition pruning, the predicates are not translated by it but evaluated in `FileIndex`, or delegated to `ExternalCatalog` (e.g. `HiveExternalCatalog` will translate to Hive's predicate for pruning).

At `FileIndex`, it is evaluated by Spark, so it is fine.

I think even here some predicates are not supported at `ExternalCatalog`, it should be orthogonal to this issue.